File: core/src/main/java/io/github/bindglam/neko/content/item/ItemBuilder.java
package io.github.bindglam.neko.content.item;

import io.github.bindglam.neko.content.feature.event.ItemStackGenerationEvent;
import io.github.bindglam.neko.platform.PlatformAdapter;
import io.github.bindglam.neko.platform.PlatformItemStack;
import io.github.bindglam.neko.utils.Constants;
import net.kyori.adventure.key.Key;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import net.kyori.adventure.text.format.TextDecoration;
import org.jetbrains.annotations.NotNull;

public final class ItemBuilder {
    public static final Key NEKO_ITEM_KEY = Key.key(Constants.MOD_ID, "item");

    private ItemBuilder() {
    }

    public static PlatformItemStack create(@NotNull Item item) {
        var itemStack = item.properties().type().createItemStack();

        Component displayName = item.properties().name();
        if (displayName == null)
            displayName = Component.translatable(item).color(NamedTextColor.WHITE);
        if (displayName.decoration(TextDecoration.ITALIC) == TextDecoration.State.NOT_SET)
            displayName = displayName.decoration(TextDecoration.ITALIC, false);
        itemStack.name(displayName);

        itemStack.lore(item.properties().lore());

        itemStack.persistentDataContainer(persistentDataContainer ->
                persistentDataContainer.set(NEKO_ITEM_KEY, PlatformAdapter.adapter().persistentDataType(String.class).orElseThrow(), item.key().asString()));

        item.featureEventBus().call(new ItemStackGenerationEvent(itemStack));

        return itemStack;
    }
}

File: core/src/main/kotlin/io/github/bindglam/neko/content/item/ItemBuilder.kt
package io.github.bindglam.neko.content.item

import io.github.bindglam.neko.utils.PLUGIN_ID
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.text.format.TextDecoration
import org.bukkit.NamespacedKey
import org.bukkit.inventory.ItemStack
import org.bukkit.persistence.PersistentDataType

object ItemBuilder {
    val NEKO_ITEM_KEY = NamespacedKey(PLUGIN_ID, "item")

    fun create(item: Item) = ItemStack(item.properties().type).apply {
        editMeta { meta ->
            meta.displayName((item.properties().name ?: Component.translatable(item).color(NamedTextColor.WHITE)).let { name ->
                if(name.decoration(TextDecoration.ITALIC) == TextDecoration.State.NOT_SET)
                    name.decoration(TextDecoration.ITALIC, false)
                else
                    name
            })
            meta.lore(item.properties().lore)

            meta.persistentDataContainer.set(NEKO_ITEM_KEY, PersistentDataType.STRING, item.key().asString())
        }
    }
}